SilKit/source/services/logging/ILoggerInternal.hpp

```diff
@@ -27,11 +27,111 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */
 
 #include "SilKitFmtFormatters.hpp"
 #include "fmt/format.h"
+#include <unordered_map>
+#include <string>
+
+
+
 
 namespace SilKit {
 namespace Services {
 namespace Logging {
 
+class LoggerMessage;
+
+
+struct ILoggerInternal : ILogger
+{
+    virtual void Log(const LoggerMessage& msg) = 0;
+    virtual void Log(const LogMsg& msg) = 0;
+};
+
+
+
+template <typename... Args>
+void Log(ILogger* logger, Level level, const char* fmt, const Args&... args);
+
+
+class LoggerMessage
+{
+public:
+    LoggerMessage(ILoggerInternal* logger, Level level)
+        : _logger(logger)
+        , _level(level)
+        , _minlevel(logger->GetLogLevel())
+    {}
+
+    LoggerMessage(ILoggerInternal* logger)
+        : _logger(logger)
+        , _level(Level::Trace)
+        , _minlevel(logger->GetLogLevel())
+    {}
+
+    LoggerMessage(ILoggerInternal* logger, const LogMsg& msg)
+        : _logger(logger)
+        , _level(msg.level)
+        , _msg(msg.payload)
+        , _keyValues(msg.keyValues)
+    {}
+
+    template <typename... Args>
+    void SetMessage(const char* fmt, const Args&... args)
+    {
+        _msg = fmt::format(fmt, args...);
+    }
+
+    void SetMessage(std::string newMsg)
+    {
+        _msg = newMsg;
+    }
```

Comment:
Use [perfect forwarding](https://en.cppreference.com/w/cpp/utility/forward) to avoid copying.

```suggestion
    template <typename T>
    void SetMessage(T&& newMsg)
    {
        _msg = std::forward<T>(newMsg);
    }
```